### PR TITLE
[BEAM-314] Add zip compression support in TextIO

### PR DESCRIPTION
**Pull request by jbonofre**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [X] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [X] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [X] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [X] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---
Add zip compression support in TextIO.

### CODE REVIEW

**Comment by davorbonaci** [comment]
R: @dhalperi

File: sdks/java/core/src/test/java/org/apache/beam/sdk/io/TextIOTest.java [hunk 1]
```diff
@@ -404,6 +406,37 @@ public void testCompressedRead() throws Exception {
 
   @Test
   @Category(NeedsRunner.class)
+  public void testZipCompressedRead() throws Exception {
```
Comment (dhalperi) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/io/TextIOTest.java]: please add a test with an empty (but valid) zip file.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/CompressedSource.java [hunk 2]
```diff
@@ -149,6 +150,26 @@ public ReadableByteChannel createDecompressingChannel(ReadableByteChannel channe
...
+      }
+
+      public ReadableByteChannel createDecompressingChannel(ReadableByteChannel channel)
+        throws IOException {
+        ZipInputStream zip = new ZipInputStream(Channels.newInputStream(channel));
+        // TODO deal with multi-entries zip
+        if (zip.getNextEntry() != null) {
+          return Channels.newChannel(zip);
+        }
+        return null;
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/CompressedSource.java]: should you instead throw an `UnsupportedOperationException` explaining the bug?

And maybe file a related JIRA?

File: sdks/java/core/src/test/java/org/apache/beam/sdk/io/TextIOTest.java [hunk 3]
```diff
@@ -404,6 +406,37 @@ public void testCompressedRead() throws Exception {
...
+
+    PrintStream writer = new PrintStream(out);
+    for (String line : lines) {
+      writer.println(line);
+      expected.add(line);
+    }
+    writer.close();
+    out.close();
+
+    Pipeline p = TestPipeline.create();
```
Comment (dhalperi) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/io/TextIOTest.java]: suggest extracting lines 411-426 in a utility function that will create a zip file. Then you can reuse it for multiple zip file tests.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/io/TextIOTest.java [hunk 4]
```diff
@@ -404,6 +406,37 @@ public void testCompressedRead() throws Exception {
 
   @Test
   @Category(NeedsRunner.class)
```
Comment (dhalperi) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/io/TextIOTest.java]: (@jbonofre no action needed, this is just copying another test and has no new issues w.r.t. this comment. -- comments below are just a heads-up)

@tgroh I'm sure you're thinking about these tests... Needs runner and also needs local filesystem to be the same system as pipeline. As we improve things, should we be (e.g.) using IOChannelUtils and such instead to create the temp files?

File: sdks/java/core/src/test/java/org/apache/beam/sdk/io/TextIOTest.java [hunk 5]
```diff
@@ -404,6 +406,37 @@ public void testCompressedRead() throws Exception {
...
+      writer.println(line);
+      expected.add(line);
+    }
+    writer.close();
+    out.close();
+
+    Pipeline p = TestPipeline.create();
+
+    TextIO.Read.Bound<String> read =
+        TextIO.Read.from(filename).withCompressionType(CompressionType.ZIP);
```
Comment (dhalperi) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/io/TextIOTest.java]: also add a test with default mode, which should auto-detect the ZIP files based on extension

File: sdks/java/core/src/test/java/org/apache/beam/sdk/io/TextIOTest.java (hunk 1 above)
Comment (jbonofre) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/io/TextIOTest.java]: Good idea, let me do it.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/io/TextIOTest.java (hunk 5 above)
Comment (jbonofre) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/io/TextIOTest.java]: Good idea, I'm doing that.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/io/TextIOTest.java (hunk 4 above)
Comment (jbonofre) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/io/TextIOTest.java]: Not sure I understand what you mean. You mean that @Category is not required ?

Comment (dhalperi) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/io/TextIOTest.java]: @jbonofre nope, no changes needed in this PR; not asking anything from you.

For now, just trying to seed a conversation about the `@Category` tag in general. Specifically, "needs runner" is not currently descriptive enough. This test wouldn't run in Cloud Dataflow and probably also not in a Flink or Spark cluster -- it relies on the local file system.

**Comment by jbonofre** [comment]
Rebased and updated. I have to figure out the expectException issue in the test.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/CompressedSource.java [hunk 6]
```diff
@@ -149,6 +150,26 @@ public ReadableByteChannel createDecompressingChannel(ReadableByteChannel channe
...
+      }
+
+      public ReadableByteChannel createDecompressingChannel(ReadableByteChannel channel)
+        throws IOException {
+        ZipInputStream zip = new ZipInputStream(Channels.newInputStream(channel));
+        // TODO deal with multi-entries zip
+        if (zip.getNextEntry() != null) {
+          return Channels.newChannel(zip);
+        }
+        throw new IllegalArgumentException("ZIP file doesn't contain any entry");
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/CompressedSource.java]: What will the behavior be today on a multi-entry zip? Will it silently produce bad data? Fail in some way?

Please comment, and then also add a test.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/io/TextIOTest.java [hunk 7]
```diff
@@ -404,6 +407,63 @@ public void testCompressedRead() throws Exception {
...
+
+    ZipOutputStream out = new ZipOutputStream(new FileOutputStream(tmpFile));
+
+    out.close();
+
+    Pipeline p = TestPipeline.create();
+
+    expectedException.expect(IllegalArgumentException.class);
+    p.apply(TextIO.Read.from(filename).withCompressionType(CompressionType.ZIP));
+
```
Comment (dhalperi) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/io/TextIOTest.java]: missing `p.run()`?

File: sdks/java/core/src/test/java/org/apache/beam/sdk/io/TextIOTest.java (hunk 4 above)
Comment (tgroh) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/io/TextIOTest.java]: Yes. `NeedsRunner` isn't really descriptive enough for what we want. This test requires a capability of Local File System Access.

Alternatively, the test could be written to access the `tempLocation` via `IOChannelUtils`

Comment (jbonofre) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/io/TextIOTest.java]: Thanks for the update guys. I now understand better ;)

**Comment by jbonofre** [comment]
Rebased and updated.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/io/TextIOTest.java [hunk 8]
```diff
@@ -412,6 +414,95 @@ public void testCompressedRead() throws Exception {
...
+    TextIO.Read.Bound<String> read =
+        TextIO.Read.from(filename).withCompressionType(CompressionType.ZIP);
+    PCollection<String> output = p.apply(read);
+
+    PAssert.that(output).containsInAnyOrder(expected);
+    p.run();
+
+    // test with auto-detect ZIP based on extension.
+    p = TestPipeline.create();
```
Comment (dhalperi) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/io/TextIOTest.java]: please make this a separate unit test -- that way they can pass or fail independently :)

File: sdks/java/core/src/test/java/org/apache/beam/sdk/io/TextIOTest.java [hunk 9]
```diff
@@ -412,6 +414,95 @@ public void testCompressedRead() throws Exception {
...
+    ZipOutputStream out = new ZipOutputStream(new FileOutputStream(tmpFile));
+
+    out.close();
+
+    Pipeline p = TestPipeline.create();
+
+    expectedException.expect(IllegalArgumentException.class);
+    p.apply(TextIO.Read.from(filename).withCompressionType(CompressionType.ZIP));
+
+    p.run();
```
Comment (dhalperi) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/io/TextIOTest.java]: Can you explain this test better? I would expect to see you looking for some sort of PipelineExecutionException

File: sdks/java/core/src/test/java/org/apache/beam/sdk/io/TextIOTest.java [hunk 10]
```diff
@@ -412,6 +414,95 @@ public void testCompressedRead() throws Exception {
...
+    out.close();
+
+    Pipeline p = TestPipeline.create();
+
+    TextIO.Read.Bound<String> read =
+        TextIO.Read.from(filename).withCompressionType(CompressionType.ZIP);
+    PCollection<String> output = p.apply(read);
+
+    PAssert.that(output).containsInAnyOrder(expected);
+    p.run();
```
Comment (dhalperi) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/io/TextIOTest.java]: I'm not sure what this is testing. Since all files are empty, I can't tell whether it read 1 of the files, 2 of the files, 3 of the files, or none of the files.

It would be good if all of the entries were non-empty and different, so that we could tell what the desired behavior was and verify that it worked against the other cases.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/io/TextIOTest.java [hunk 11]
```diff
@@ -412,6 +414,95 @@ public void testCompressedRead() throws Exception {
...
+    read = TextIO.Read.from(filename);
+    output = p.apply(read);
+
+    PAssert.that(output).containsInAnyOrder(expected);
+    p.run();
+  }
+
+  @Test
+  @Category(NeedsRunner.class)
+  public void testZipCompressedReadWithEmptyFile() throws Exception {
```
Comment (dhalperi) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/io/TextIOTest.java]: Can you also add a test where the zip file is valid, has a single entry, and that entry is empty? In other words, a zip of an empty file. Seems like another corner case to handle!

File: sdks/java/core/src/test/java/org/apache/beam/sdk/io/TextIOTest.java (hunk 9 above)
Comment (dhalperi) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/io/TextIOTest.java]: (@tgroh , suggestions for handling the exception?)

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/CompressedSource.java [hunk 12]
```diff
@@ -149,6 +150,29 @@ public ReadableByteChannel createDecompressingChannel(ReadableByteChannel channe
         return Channels.newChannel(
             new BZip2CompressorInputStream(Channels.newInputStream(channel)));
       }
+    },
+
+    /**
+     * Reads a byte channel assuming it is compressed with zip.
+     */
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/CompressedSource.java]: can you add more to the Javadoc here?

```
Only zip files that have a single entry are supported.
```

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/CompressedSource.java [hunk 13]
```diff
@@ -149,6 +150,29 @@ public ReadableByteChannel createDecompressingChannel(ReadableByteChannel channe
...
+      }
+
+      public ReadableByteChannel createDecompressingChannel(ReadableByteChannel channel)
+        throws IOException {
+        ZipInputStream zip = new ZipInputStream(Channels.newInputStream(channel));
+        if (zip.getNextEntry() != null) {
+          if (zip.getNextEntry() != null) {
+            throw new IllegalArgumentException("Multi-entries ZIP file is not supported");
+          } else {
+            return Channels.newChannel(zip);
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/CompressedSource.java]: does this work? Based on how you wrote the code, I would expect that `zip.getNextEntry()` actually mutates the ZipInputStream in some way. Is it still valid to open a stream from the zip after calling getNextEntry twice?

File: sdks/java/core/src/test/java/org/apache/beam/sdk/io/TextIOTest.java (hunk 10 above)
Comment (dhalperi) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/io/TextIOTest.java]: Based on your code in the actual zip decompresser, I expect this to fail. But the test itself does not expect failure.

**Comment by dhalperi** [comment]
Hi JB,

This is looking pretty good!

But I have some questions about the tests. Specifically, since we mostly test empty files it seems tough to validate that the decompressor does exactly what we expect.

I've added some suggestions for improvements.

Thanks,
Dan

File: sdks/java/core/src/test/java/org/apache/beam/sdk/io/TextIOTest.java (hunk 9 above)
Comment (tgroh) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/io/TextIOTest.java]: We don't wrap exceptions thrown from Source methods in `UserCodeExceptions`, so they won't get rewrapped in `PipelineExecutionExceptions`.

I would put the `expectedException` right before the line that is expected to throw, however.

**Comment by jbonofre** [comment]
Updated. I added explanations on each test.
However, I have two points:
1. Using `ZipInputStream` there's not way to get the number of entries. So, unfortunately, I don't see a good way to raise an exception when the zip stream contains multiple entries.
2. In the `testZipCompressedReadWithEmptyFile` test, I got the right `IllegalArgumentException` but wrapped in a `RuntimeException`. That's why I'm doing `expectedException.expect(RuntimeException.class)` instead of `expectedException.expect(IllegalArgumentException.class)`. Any idea why ?

File: sdks/java/core/src/test/java/org/apache/beam/sdk/io/TextIOTest.java [hunk 14]
```diff
@@ -410,6 +412,158 @@ public void testCompressedRead() throws Exception {
...
+    out.close();
+
+    Pipeline p = TestPipeline.create();
+
+    TextIO.Read.Bound<String> read =
+        TextIO.Read.from(filename).withCompressionType(CompressionType.ZIP);
+    PCollection<String> output = p.apply(read);
+
+    PAssert.that(output).empty();
+    p.run();
```
Comment (dhalperi) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/io/TextIOTest.java]: I'd really like to find a way that this case either concats all the files or throws an exception. The current behavior is effectively silent data loss, the worst possible case!

One way to handle this could be a utility InputStream class that wraps `ZipInputStream` and under-the-hood concats all the different entry streams. This is probably the base case.

A second possibility is that once the input stream hits EOF, we check for a next entry and only then throw an exception. But this is less desirable as we don't fail until pretty late.

Can you look into it?

Comment (jbonofre) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/io/TextIOTest.java]: I like your idea of wrapping the stream. Let me figure it out.

**Comment by jbonofre** [comment]
I updated the PR to use another approach: I'm using directly the `ZipInputStream` with considering the entries. It allows user to read multi-entries zip file as single entry one. I also updated the tests to show the "new" behavior.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/io/TextIOTest.java [hunk 15]
```diff
@@ -410,6 +412,170 @@ public void testCompressedRead() throws Exception {
...
+    PCollection<String> output = p.apply(TextIO.Read.from(filename).withCompressionType
+        (CompressionType.ZIP));
+    PAssert.that(output).empty();
+
+    p.run();
+  }
+
+  /**
+   * Read a ZIP compressed file with multiple entries. Only the first entry is actually read.
+   * We expect an empty PCollection.
```
Comment (dhalperi) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/io/TextIOTest.java]: update the javadoc to be correct.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/io/TextIOTest.java [hunk 16]
```diff
@@ -410,6 +412,170 @@ public void testCompressedRead() throws Exception {
...
+
+    TextIO.Read.Bound<String> read = TextIO.Read.from(filename);
+    PCollection<String> output = p.apply(read);
+
+    PAssert.that(output).containsInAnyOrder(expected);
+    p.run();
+  }
+
+  /**
+   * Read a ZIP compressed empty file. We expect an exception.
```
Comment (dhalperi) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/io/TextIOTest.java]: update the javadoc?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/CompressedSource.java [hunk 17]
```diff
@@ -149,6 +150,25 @@ public ReadableByteChannel createDecompressingChannel(ReadableByteChannel channe
         return Channels.newChannel(
             new BZip2CompressorInputStream(Channels.newInputStream(channel)));
       }
+    },
+
+    /**
+     * Reads a byte channel assuming it is compressed with zip.
+     *
+     * Warning: only zip files that have exactly one entry (aka are ZIPs of a single file) are
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/CompressedSource.java]: update the javadoc?

File: sdks/java/core/src/test/java/org/apache/beam/sdk/io/TextIOTest.java [hunk 18]
```diff
@@ -410,6 +412,170 @@ public void testCompressedRead() throws Exception {
...
+  @Category(NeedsRunner.class)
+  public void testZipCompressedReadWithMultiEntriesFile() throws Exception {
+    File tmpFile = tmpFolder.newFile("testZipCompressedAutoDetected.zip");
+    String filename = tmpFile.getPath();
+
+    List<String> expected = new ArrayList<>();
+
+    ZipOutputStream out = new ZipOutputStream(new FileOutputStream(tmpFile));
+
+    PrintStream writer = new PrintStream(out);
```
Comment (dhalperi) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/io/TextIOTest.java]: I think you need to close writer before you call `out.closeEntry`, and reopen a new `writer` for every new entry.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/io/TextIOTest.java (hunk 3 above)
Comment (dhalperi) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/io/TextIOTest.java]: still think it would be a good idea to try to refactor the tests to reduce code reuse, if you can take a stab

**Comment by dhalperi** [comment]
Thanks JB -- looking great.

* Please ensure that the multi-entry test is actually passing by fixing the way the writers are used.
* Fix the javadoc with new semantics.

Other than that, I'd love to see less code reuse in the tests. This is considered a strong recommendation, but I won't withhold and LGTM on that basis ;).

**Comment by jbonofre** [comment]
My bad about the javadoc. I will fix that. I'm also fixing the multi-entries test. Then I will refactore a bit to use a share method for zip file creation ;)

**Comment by jbonofre** [comment]
PR rebased and updated based on Dan's comments.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/io/TextIOTest.java [hunk 19]
```diff
@@ -410,6 +412,148 @@ public void testCompressedRead() throws Exception {
...
+    File tmpFile;
+    if (filename != null) {
+      tmpFile = tmpFolder.newFile(filename);
+    } else {
+      tmpFile = tmpFolder.newFile();
+    }
+    String tmpFileName = tmpFile.getPath();
+
+    ZipOutputStream out = new ZipOutputStream(new FileOutputStream(tmpFile));
+    PrintStream writer = new PrintStream(out);
```
Comment (dhalperi) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/io/TextIOTest.java]: Please use

```java
PrintStream writer = new PrintStream(out, true /* auto-flush on write */);
```

this way we can be sure that the `PrintStream` itself does not buffer any data. See [javadoc](https://docs.oracle.com/javase/7/docs/api/java/io/PrintStream.html#PrintStream(java.io.OutputStream,%20boolean)),

File: sdks/java/core/src/test/java/org/apache/beam/sdk/io/TextIOTest.java [hunk 20]
```diff
@@ -410,6 +412,148 @@ public void testCompressedRead() throws Exception {
...
+  /**
+   * Create a zip file with the given lines.
+   *
+   * @param expected A list of expected lines, populated in the zip file.
+   * @param filename Optionally zip file name (can be null).
+   * @param fieldsEntries Fields to write in zip entries.
+   * @return The zip filename.
+   * @throws Exception In case of a failure during zip file creation.
+   */
+  private String createZipFile(List<String> expected, String filename, String[] ... fieldsEntries)
```
Comment (dhalperi) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/io/TextIOTest.java]: Mark any nullable parameters as `@Nullable`

File: sdks/java/core/src/test/java/org/apache/beam/sdk/io/TextIOTest.java [hunk 21]
```diff
@@ -410,6 +412,148 @@ public void testCompressedRead() throws Exception {
...
+    p.run();
+  }
+
+  /**
+   * Read a ZIP compressed empty file. We expect an empty PCollection.
+   */
+  @Test
+  @Category(NeedsRunner.class)
+  public void testZipCompressedReadWithEmptyFile() throws Exception {
+    String filename = createZipFile(null, null);
```
Comment (dhalperi) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/io/TextIOTest.java]: believe the first parameter here and below should be `new ArrayList<>()` -- that parameter is not `@Nullable` (even though because there is no data in the file it does not crash.

**Comment by dhalperi** [comment]
Looking great! Only trivial changes left.

**Comment by jbonofre** [comment]
Rebased and updated according to Dan's comments.

**Comment by jbonofre** [comment]
There are tests failure in the DirectRunner due to the latest changes and rebase. I will fix that.

**Comment by dhalperi** [comment]
Hi JB!

One small request:
* Please don't rebase every time. It's much easier to review if you simply add new incremental CLs to the existing pull request, as this way I can see what changed ;) (This is mentioned briefly in the [contribution guide](http://beam.incubator.apache.org/contribution-guide/#code-review-and-revision), but is not obvious and may not be standard practice everywhere).

I checked out your PR and tried to get the tests to pass -- but I was never able to make both the single-file and multi-file tests pass. I think that right now the ZipInputStream does not automatically concat all file contents without more effort. Might need to make a new class that wraps the ZipInputStream and manually calls `getNextEntry` whenever the current entry reaches EOF.

**Comment by jbonofre** [comment]
Hi Dan,

Thanks for the update and ok for the rebase (sorry about that). I'm checking and fixing the issue.

**Comment by jbonofre** [comment]
Working on a ZipInputStream wrapper (which call getNextEntry() behind the hood). Update soon.

**Comment by jbonofre** [comment]
@dhalperi It should be OK now.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/CompressedSource.java [hunk 22]
```diff
@@ -164,12 +166,40 @@ public boolean matches(String fileName) {
...
+    public class FullZipInputStream extends ZipInputStream {
+
+      private ZipEntry currentEntry;
+
+      public FullZipInputStream(InputStream is) throws IOException {
+        super(is);
+        currentEntry = getNextEntry();
+      }
+
+      public int read(byte[] b, int off, int len) throws IOException {
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/CompressedSource.java]: there are a lot of variants of `read` in an `InputStream`. Is it obvious that this is the only variant you need to override? (May well be, I just don't know).

Comment (jbonofre) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/CompressedSource.java]: Yes, it's the only read using in the IOChannel.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/CompressedSource.java [hunk 23]
```diff
@@ -149,9 +152,54 @@ public ReadableByteChannel createDecompressingChannel(ReadableByteChannel channe
...
+        int result = super.read(b, off, len);
+        if (result == -1) {
+          currentEntry = getNextEntry();
+          if (currentEntry == null) {
+            return -1;
+          }
+          result = super.read(b, off, len);
+          return result;
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/CompressedSource.java]: what if there are two empty entries in a row?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/CompressedSource.java (hunk 22 above)
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/CompressedSource.java]: I do think you should override all applicable functions from InputStream -- implementation of IOChannel might change and/or this stream might be used in a different way.

Comment (lukecwik) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/CompressedSource.java]: This is the only read that has to be implemented, all the others are just wrappers for this read method.

int read() is the only one that you may want to handle specially since its very inefficient if its not implemented but its really inefficient for people to use so having an efficient implementation is useful.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/CompressedSource.java (hunk 23 above)
Comment (lukecwik) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/CompressedSource.java]: Even if there is only one entry and its empty, super.read will return -1 representing the end of that entry which will be returned to the caller and that will prematurely end the stream ignoring any remaining zip entries.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/CompressedSource.java [hunk 24]
```diff
@@ -149,9 +152,54 @@ public ReadableByteChannel createDecompressingChannel(ReadableByteChannel channe
...
+        throws IOException {
+        FullZipInputStream zip = new FullZipInputStream(Channels.newInputStream(channel));
+        return Channels.newChannel(zip);
+      }
     };
 
     /**
+     * Extend of {@link ZipInputStream} to automatically read all entries in the zip.
+     */
+    public class FullZipInputStream extends ZipInputStream {
```
Comment (lukecwik) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/CompressedSource.java]: Please extend AbstractInputStream in which you delegate to an internal member variable representing the zip input stream. This prevents the exposure of methods on ZipInputStream.

Also, make this class private and state. Users don't need to have visibility of this implementation.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/io/TextIOTest.java [hunk 25]
```diff
@@ -419,6 +423,150 @@ public void testCompressedRead() throws Exception {
...
+  /**
+   * Create a zip file with the given lines.
+   *
+   * @param expected A list of expected lines, populated in the zip file.
+   * @param filename Optionally zip file name (can be null).
+   * @param fieldsEntries Fields to write in zip entries.
+   * @return The zip filename.
+   * @throws Exception In case of a failure during zip file creation.
+   */
+  private String createZipFile(List<String> expected, @Nullable String filename, String[]
```
Comment (lukecwik) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/io/TextIOTest.java]: formatting?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/CompressedSource.java (hunk 24 above)
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/CompressedSource.java]: s/state/static/

Comment (jbonofre) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/CompressedSource.java]: Thanks for the update guys. I'm updating accordingly.

Comment (jbonofre) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/CompressedSource.java]: AbstractInputStream doesn't exist in the JDK. Which one are you talking about ? Thanks !

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/CompressedSource.java (hunk 23 above)
Comment (jbonofre) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/CompressedSource.java]: I tested the following cases:
1. zip with two entries (non empty)
2. zip with the first entry empty and the second one with data
3. zip with the first entry with data and the second one empty

It works fine:

```
    String filename = createZipFile(new ArrayList<String>(), null, new String[]{}, new
        String[]{"three", "four"});

    FullZipInputStream zipInputStream = new FullZipInputStream(new FileInputStream(new File
        (filename)));
    BufferedReader reader = new BufferedReader(new InputStreamReader(zipInputStream));
    String line;
    while ((line = reader.readLine()) != null) {
      System.out.println(line);
    }
```

Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/CompressedSource.java]: Hi JB, Here's a test that fails that exhibits the problem I'm worried about.

```java
  /**
   * Read a ZIP compressed file containing data, multiple empty entries, and then more data. We
   * expect just the data back.
   */
  @Test
  @Category(NeedsRunner.class)
  public void testZipCompressedReadWithComplexEmptyAndPresentEntries() throws Exception {
    String filename = createZipFile(
        new ArrayList<String>(),
        null,
        new String[] {"cat"},
        new String[] {},
        new String[] {},
        new String[] {"dog"});
    List<String> expected = ImmutableList.of("cat", "dog");

    Pipeline p = TestPipeline.create();

    PCollection<String> output =
        p.apply(TextIO.Read.from(filename).withCompressionType(CompressionType.ZIP));
    PAssert.that(output).containsInAnyOrder(expected);

    p.run();
  }
```

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/CompressedSource.java (hunk 24 above)
Comment (lukecwik) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/CompressedSource.java]: I meant to have said InputStream since its the abstract base class and not an interface instead of AbstractInputStream.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/CompressedSource.java (hunk 23 above)
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/CompressedSource.java]: I admit two empty files is relatively unlikely, though totally possible. But this is not just some weird corner case--there are other execution paths in which it is possible for the stream to read exactly to the end of a file but not hit EOF (so `-1` is not returned yet), and then it will act like an empty file for the purposes of the next call into the function.

**Comment by dhalperi** [comment]
JB this is awesome. One small fix to get the multiple-eof-in-a-row case fixed and let's merge it. A big accomplishment!

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/CompressedSource.java (hunk 23 above)
Comment (jbonofre) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/CompressedSource.java]: Catcha, fixing that. Thanks !

**Comment by jbonofre** [comment]
Great feedback guys ! Much appreciated. Fixing the multi-entries. Thanks !

**Comment by jbonofre** [comment]
Fixed the `read()` methods and now extends `InputStream` wrapping the `ZipInputStream` instead of extending it.

**Comment by dhalperi** [comment]
LGTM. Jenkins flakes on network errors, but Travis is fine. Merging.
